### ISSUE TITLE
desi_group_spectra incorrectly assumes MPI

### ISSUE THREAD

**Issue by sbailey** [desihub/desispec#1357]
desi_group_spectra uses desispec.parallel.use_mpi to try to guess if MPI is being used, which defaults to True if you are on a batch node even if you called desi_group_spectra without an srun prefix.  The workaround is to use `srun -n 1 desi_group_spectra ...` when running on a batch node, but that incurs extra overhead/fragility and limits flexibility for using things like gnu parallel.

Options:
  * update use_mpi to have additional checks for SLURM environment variables that won't exist if it wasn't called with MPI, but don't automatically break non-slurm systems.  e.g. if SLURM_NODEID is set but SLURM_NTASKS is *not* set, it means you are on an interactive node but running something that wasn't prefixed with srun.  If SLURM_NTASKS *is* set, then you are running under srun.  But if even SLURM_NODEID isn't set, then you might be on a non-slurm system and all bets are off.
  * update desi_group_spectra to be like other scripts like desi_proc which require a specific `--mpi` opt-in (or like rrdesi_mpi a separate mpi-specific script).  This comes with the risk of using srun but forgetting to use --mpi (or desi_group_spectra_mpi).
  * if anyone else has a better recipe for a script introspecting its environment to determine if it was called with an srun/aprun/mpirun prefix, let us know...
    * Note: simply trying "from mpi4py import MPI" doesn't work because on NERSC login nodes that crashes so hard that python exception handling can't catch it.  We need something that can know whether it is safe to try to import mpi4py...

**Comment by lgarrison** [comment]
I stumbled across this issue and spent a few minutes thinking about it.  One hacky solution that seems to work is to spin off the `import mpi4py.MPI` as a subprocess, so that a shared-library error doesn't cause the whole house of cards to collapse.  This could look like:

```python
from multiprocessing import Process

def import_mpi4py():
    # could probably suppress scary errors here with os.dup2() tricks
    from mpi4py import MPI

def use_mpi():
    p = Process(target=import_mpi4py)
    p.start()
    p.join()
    return p.exitcode == 0

if __name__ == '__main__':
    print(use_mpi())
```

Try with `python use_mpi.py` and `srun -n1 use_mpi.py`; seems to give the right answers.  Of course, this is a self-contained test, and I could totally imagine that there would be problems with forking an actual desispec process.  Then again, I see that desispec uses multiprocessing elsewhere, so maybe it's safe?  Definitely hacky, regardless!

**Comment by sbailey** [comment]
Interesting idea.  MPI is famously grumpy about not playing nicely with other parallelism models and this smells dangerous.  Experts disagree about whether the standard even permits MPI ranks to fork (vs. just being fragilely lucky if it doesn't crash on you), but even those who do think it allows forking agree that the forked processes are not supposed to do anything MPI-specific (like importing mpi4py).  e.g. imagine a 1000-rank job which has 1000 new processes spawned that connect to MPI and then exit before the original parent processes connect to MPI ... you could imagine MPI not being prepared to handle that gracefully.

For related discussion and links, see this openmpi FAQ entry: https://icl.cs.utk.edu/open-mpi/faq/?category=openfabrics#ofa-fork

**Comment by lgarrison** [comment]
Yeah, I agree, it's a mess; I've run into the "MPI processes should not fork" issue before.  My guess it that it would be safe (or at least safer) as long as the forking is done before MPI initialized in the parent process, as in this example.  But I agree having each of 1000 MPI ranks launch its own subprocess just to make this simple determination is a recipe for disaster.
